src/tiered/tiered_handle.c

```diff
@@ -228,6 +228,58 @@ __tiered_create_local(WT_SESSION_IMPL *session, WT_TIERED *tiered)
     return (ret);
 }
 
+/*
+ * __tiered_flush_older_objects --
+ *     Check earlier objects and see if flush work units for earlier objects need to be pushed onto
+ *     the work queue. This can happen if the system crashed during a flush tier.
+ */
+static int
+__tiered_flush_older_objects(WT_SESSION_IMPL *session, WT_TIERED *tiered)
+{
+    WT_CONFIG_ITEM cval;
+    WT_DECL_RET;
+    uint32_t i;
+    const char *obj_name, *obj_uri, *obj_val;
+    bool exist;
+
+    /* If this is a new table and there are no objects there is nothing to do. */
+    if (tiered->current_id == 0)
+        return (0);
+
+    /*
+     * Work our way backwards through all earlier objects and look at each object's flush_time
+     * configuration setting. If it is zero that means the object is not flushed and a work unit
+     * will be pushed for that object.
+     */
+    obj_name = obj_uri = obj_val = NULL;
+    exist = false;
+    for (i = tiered->current_id - 1; i > 0; --i) {
+        WT_ERR(__wt_tiered_name(session, &tiered->iface, i, WT_TIERED_NAME_OBJECT, &obj_uri));
+        obj_name = obj_uri;
+        WT_PREFIX_SKIP_REQUIRED(session, obj_name, "object:");
+        WT_ERR(__wt_fs_exist(session, obj_name, &exist));
+        /*
+         * We only need to worry about objects that exist locally. As soon as we find an object that
+         * does not exist, that means that it, and all earlier objects, have been flushed.
+         */
```

Comment:
I want to be sure this comment is true.  Is there any chance that objects may get removed out of order?  For example, assume (because of network errors), that I start with 2 objects (.0001 and .0002) still in the file system, not flushed.  So I queue them both to be flushed.  .0002 finishes - is it possible that the local copy of 0002 is removed?  If so, and I crash after that, then doing this loop on the next restart, I'll stop at 0002, and won't ever flush 0001.  Or is there something to prevent that scenario?

Reply:
Thanks for this. I think the only way it could be possible is in this code and I'll reverse the direction of the loop. We know (and I should have used) the `oldest` tiered object number instead of `1`.

Every `flush_tier` call must wait for the previous one to complete, either before returning to the user, or before the next one begins. That means waiting for a subsequent checkpoint and then flushing all the objects. So there should be, at most, one object possibly outstanding. In case there is an edge case of two, I will reverse the order of the loop and make sure earlier objects get resolved first.